lib/srv/regular/sshserver.go

```diff
@@ -1096,40 +1112,255 @@ func (s *Server) getServerResource() (types.Resource, error) {
+func (s *Server) getDirectTCPIPForwardDialer(scx *srv.ServerContext) (sshutils.TCPIPForwardDialer, error) {
+	if d, ok := scx.Parent().GetDirectTCPIPForwardDialer(); ok {
 		return d, nil
 	}
 
-	dialerConn, listenerConn, err := uds.NewSocketpair(uds.SocketTypeDatagram)
+	proc, err := s.startForwardingSubprocess(scx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// set up a dial function that sends the address + fd as a unix datagram message. the forwarder subprocess
+	// interprets all such messages in this way, and will dial the specified address and proxy all traffic
+	// to the desired endpoint.
+	dialer := func(addr string) (net.Conn, error) {
+		local, remote, err := uds.NewSocketpair(uds.SocketTypeStream)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		defer remote.Close()
+
+		remoteFD, err := remote.File()
+		if err != nil {
+			local.Close()
+			return nil, trace.Wrap(err)
+		}
+		defer remoteFD.Close()
+
+		_, _, err = proc.Conn.WriteWithFDs([]byte(addr), []*os.File{remoteFD})
+		if err != nil {
+			local.Close()
+			return nil, trace.Wrap(err)
+		}
+
+		return local, nil
+	}
+
+	// try to register with the parent context.
+	if other, ok := scx.Parent().TrySetDirectTCPIPForwardDialer(dialer); !ok {
+		// another forwarder was concurrently created. this isn't actually a problem, multiple forwarders
+		// being registered is harmless, but it does result in slightly higher resource utilization, so its
+		// preferable to use the existing forwarder and close ours in the background.
+		go proc.Close()
+		return other, nil
+	}
+
+	// successfully registered this dialer, add closer to context.
+	scx.Parent().AddCloser(proc)
+
+	return dialer, nil
+}
+
+// listenTCPIP creates a new listener in the forwarding process.
+func (s *Server) listenTCPIP(scx *srv.ServerContext, addr string) (*net.TCPListener, error) {
+	proc, err := s.getTCPIPForwardProcess(scx)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
-	defer listenerConn.Close()
 
-	listenerFD, err := listenerConn.File()
+	// Create a socket to receive new connections on.
+	localConn, remoteConn, err := uds.NewSocketpair(uds.SocketTypeDatagram)
 	if err != nil {
-		dialerConn.Close()
 		return nil, trace.Wrap(err)
 	}
-	defer listenerFD.Close()
+	defer localConn.Close()
+	defer remoteConn.Close()
+	remoteFD, err := remoteConn.File()
+	if err != nil {
+		localConn.Close()
+		return nil, trace.Wrap(err)
+	}
+	defer remoteFD.Close()
+	_, _, err = proc.Conn.WriteWithFDs([]byte(addr), []*os.File{remoteFD})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// The forwarding process may have chosen its own port, so we need to get the
+	// new listen address.
+	fileCh := make(chan *os.File)
+	// Read addr in another goroutine so we can cancel it if the forwarding process
+	// stops.
+	go func() {
+		defer close(fileCh)
+
+		fbuf := make([]*os.File, 1)
+		_, fn, err := localConn.ReadWithFDs(nil, fbuf)
+		if err != nil || fn == 0 {
+			fileCh <- nil
+		}
```

Comment:
I'm not entirely sure one can `recvmsg()` and get data _and_ an error, but we shouldn't risk it:

```suggestion
		if _, fn, _ := localConn.ReadWithFDs(nil, fbuf); fn == 0 {
			fileCh <- nil
		}
```